js/src/data/utils.test.js

```diff
@@ -263,3 +264,34 @@ describe( 'mapReportFieldsToPerformance', () => {
 		} );
 	} );
 } );
+
+describe( 'getConnectedJetpackInfo', () => {
+	it( 'When jetpack in not connected, should return an empty string', () => {
+		expect( getConnectedJetpackInfo( { active: 'no' } ) ).toBe( '' );
+	} );
+
+	describe( 'For connected jetpack', () => {
+		it( 'When current user is the jetpack owner, should return email', () => {
+			const jetpack = {
+				active: 'yes',
+				owner: 'yes',
+				email: 'i_am_jetpack_owner@example.com',
+			};
+
+			expect( getConnectedJetpackInfo( jetpack ) ).toBe( jetpack.email );
+		} );
+
+		it( 'When current user is not the jetpack owner, should return a successfully connected info', () => {
+			const jetpack = {
+				active: 'yes',
+				owner: 'no',
+				email: 'i_am_another_admin_user@example.com',
+			};
+
+			const info = getConnectedJetpackInfo( jetpack );
+
+			expect( info ).not.toBe( jetpack.email );
+			expect( info ).toMatchSnapshot();
```

Comment:
❓  Since the phrase `"Successfully connected through Jetpack"` is a requirement as defined by Jill in https://github.com/woocommerce/google-listings-and-ads/issues/624#issuecomment-883829052, I would code it into the test. One `expect` instead of two, and we don't need the `js/src/data/__snapshots__/utils.test.js.snap` file. What do you think?

```suggestion
			expect( info ).toBe( 'Successfully connected through Jetpack' );
```

Reply:
The main concern of this test is to verify whether the function makes logical condition based on `owner`, and the returned textual content is not the primary target of this test. Using snapshot can still verify whether the text content is correct. When the text change happened, the snapshot could alert developers as well and could be easily updated if the change is confirmed. Furthermore, it doesn't need to maintain another text copy in test case.

> One `expect` instead of two, and we don't need the `js/src/data/__snapshots__/utils.test.js.snap` file. What do you think?

However, using only the snapshot to verify the condition would be a bit ambiguous in expressing the test intention, so an additional `expect( info ).not.toBe( jetpack.email );` is used to enhance the expressiveness. Just like the `checkErrors` test, where the snapshot is verified along with an `expect( errors ).toHaveProperty( 'shipping_rate' );` to strengthen the expression of test intention.

https://github.com/woocommerce/google-listings-and-ads/blob/20247e60a491f2546bcaee5862d30f678b01557f/js/src/components/free-listings/configure-product-listings/checkErrors.test.js#L68-L73

---

I have to say that this test case is not precise enough, either with snapshot or with `.toBe()` assertion. When I developed it, I thought the `getConnectedJetpackInfo` function was so straightforward that it was almost impossible to get it wrong, so I'm probably okay with a little less precision.

But to make it precise, then I would change it to a pure function like this

```js
// Change function to
function getConnectedJetpackInfo( jetpack, nonOwnerInfo ) {
	// ... omitted

	if ( jetpack.owner === 'yes' ) {
		return jetpack.email;
	}

	// No available owner info. Show `nonOwnerInfo` instead.
	return nonOwnerInfo;
}
```

And assert it by

```js
// Assert it by
it( 'When current user is not the jetpack owner, should return `nonOwnerInfo` param instead', () => {
	const jetpack = {
		active: 'yes',
		owner: 'no',
		email: 'i_am_another_admin_user@example.com',
	};
	const nonOwnerInfo = {}; // Use object to assert the same reference

	const info = getConnectedJetpackInfo( jetpack, nonOwnerInfo );

	expect( info ).toBe( nonOwnerInfo );
} );
```

Based on the premise of a very simple function, that would be a bit overkill for me and need more codes when used. But it ensures the accuracy of the test. So, in terms of overall balance, I chose the current implementation when I was developing it. Which would you prefer?

Reply:
> ... to verify whether the function makes logical condition based on `owner`, and the returned textual content is not the primary target of this test.

When I first see the function and its JSDoc, I thought the responsibility of the function is to make the logical condition, plus having the string `'Successfully connected through Jetpack'` in one place, so that other places calling the function will get the same string. That's why I thought it make sense to test for the string. What do you think on this?

> Furthermore, it doesn't need to maintain another text copy in test case.

Having the text copy in test case is common and expected, right? I mean, if there is a requirement that says "the component / function / application should display 'Hello World'", the test for it would need to assert the existence of the 'Hello World' text.

> Which would you prefer?

The one you presented here:

```
	const info = getConnectedJetpackInfo( jetpack, nonOwnerInfo );
```

... yeah, it does look overkill. Also, it would require the `nonOwnerInfo` (i.e. the string `'Successfully connected through Jetpack'`) to be placed elsewhere and to be passed into the function call, which feels like not so ideal in terms of developer experience (which is why I thought good to put the string in the function).

So comparing the code you have in this PR and the one you presented above, the one in this PR in good. 😄

Reply:
> That's why I thought it make sense to test for the string. What do you think on this?

For this case, the wording of expression is not the main focus of the test. I thought snapshot is an appropriate usage, and it will test whether the string is matched as well.

There are many similar contexts in [checkErrors.test.js](https://github.com/woocommerce/google-listings-and-ads/blob/20247e60a491f2546bcaee5862d30f678b01557f/js/src/components/free-listings/configure-product-listings/checkErrors.test.js). After viewing every `toMatchSnapshot()` usages, maybe it could show more sense why I adopted this method. But If devs thought that every time they read these tests, they have to open the function implementation or snapshot files to confirm what the string exactly is, it really affects the developer experience. Then I would be happy to open another PR to have these strings within test cases directly.

> Having the text copy in test case is common and expected, right?

If a function is like the below, having the string copy within test case is common and expected.

```js
function detectSocialNetwork( url ) {
	if ( matchTwitter( url ) ) {
		return 'twitter';
	}
	if ( matchFacebook( url ) ) {
		return 'facebook';
	}
	if ( matchInstagram( url ) ) {
		return 'instagram';
	}
	return null;
}
```

> I mean, if there is a requirement that says "the component / function / application should display 'Hello World'", the test for it would need to assert the existence of the 'Hello World' text.

I think it depends on the text is actually displayed by a component, function, or application. And how the "display" is to be defined. It would bring different practices.